packages/api/src/car.js

```diff
@@ -19,44 +23,90 @@ export async function carHead(request, env, ctx) {
-    // }
-    res = new Response(res.body, { ...res, headers })
-    ctx.waitUntil(cache.put(request, res.clone()))
+
+  if (res) {
+    return res
+  }
+
+  const {
+    params: { cid },
+  } = request
+  // gateway does not support `carversion` yet.
+  // using it now means we can skip the cache if it is supported in the future
+  const url = new URL(`/api/v0/dag/export?arg=${cid}&carversion=1`, GATEWAY)
+  console.log(url.toString())
+  res = await fetch(url, { method: 'POST' })
+  if (!res.ok) {
+    // bail early. dont cache errors.
+    return res
+  }
+  // TODO: these headers should be upstreamed to ipfs impls
+  // without the content-disposition, firefox describes them as DMS files.
+  const headers = {
+    'Cache-Control': 'public, max-age=31536000',  // max max-age is 1 year
+    'Content-Type': 'application/car',
+    'Content-Disposition': `attachment; filename="${cid}.car"`,
   }
+  // // compress if asked for? is it worth it?
+  // if (request.headers.get('Accept-Encoding').match('gzip')) {
+  //   headers['Content-Encoding'] = 'gzip'
+  // }
+  res = new Response(res.body, { ...res, headers })
+  ctx.waitUntil(cache.put(request, res.clone()))
+
   return res
 }
 
 /**
  * @param {import('./user').AuthenticatedRequest} request
+ * @param {import('./env').Env} env
  */
-export async function carPost(request, env, ctx) {
-  const { user, authToken } = request.auth
-  console.log('User ID', user._id)
-  console.log('Auth Token ID', authToken && authToken._id)
-  return new JSONResponse({})
+export async function carPost(request, env) {
+  const { authToken } = request.auth
+
+  const { headers } = request
+  const contentType = headers.get('content-type')
+
+  if (!contentType.includes('application/car')) {
+    return new Response(`${request.method} /car did not receive a car content type`, { status: 400 })
+  }
+
+  let name = headers.get('name')
+  if (!name || typeof name !== 'string') {
+    // throw new Error('invalid name') // Is it required?
+    name = 'testname'
+  }
+
+  const blob = await request.blob()
+
+  // Ensure car blob.type is set; it is used by the cluster client to set the foramt=car flag on the /add call.
+  const content = blob.slice(0, blob.size, 'application/car')
+
+  const { cid } = await cluster.add(content, {
+    // When >2.5MB, use local add, because waiting for blocks to be sent to
+    // other cluster nodes can take a long time. Replication to other nodes
+    // will be done async by bitswap instead.
+    local: blob.size > LOCAL_ADD_THRESHOLD,
+  })
+
+  // Store in DB
+  await env.db.query(gql`
+    mutation importCar($data: importCarInput!) {
+      importCar(data: $data) {
+        _id
+      }
+    }
+  `, {
+    data: { 
+      authToken,
+      cid,
+      name
+      // dagSize: undefined // TODO: should we default to chunk car behavior?
+    }
+  })
+
+  // TODO: Improve response type with pin information
+  return new JSONResponse({ ok: true, value: { cid, name } })
```

Comment:
I'm not sure what the benefit of the `{ ok, value }` pattern is. We end up with users having to check `Response.ok` and also `data.ok`...seems like a duplication.

Maybe @hugomrdias or @Gozala know the reasons?

Reply:
Yes, this was related to my comment in your PR: https://github.com/web3-storage/web3.storage/pull/30#discussion_r664653009